src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,855 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2019, Fraunhofer SIT, Infineon Technologies AG, Intel Corporation
+ * All rights reserved.
+ ******************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <dlfcn.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <netinet/in.h>
+#include "tss2_tcti_libtpms.h"
+
+#include "tcti-libtpms.h"
+#include "tcti-common.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * libtpms API calls need to be wrapped. We set the current active TCTI module
+ * for this thread. This is needed because libtpms may call callbacks and these
+ * need to know which TCTI context they have to operate on.
+ *
+ * This macro assumes that int ret is declared. Jumps to fail_label on error. In
+ * this case, rc contains the respective error code.
+ */
+#define LIBTPMS_API_CALL(fail_label, tcti_libtpms, function, ...) \
+    current_tcti_libtpms = tcti_libtpms; \
+    ret = tcti_libtpms->function(__VA_ARGS__); \
+    if (ret != TPM_SUCCESS) { \
+        LOG_ERROR("libtpms function " #function "() failed with return code 0x%" PRIx32, ret); \
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE; \
+        goto fail_label; \
+    } \
+    current_tcti_libtpms = NULL;
+
+__thread TSS2_TCTI_LIBTPMS_CONTEXT *current_tcti_libtpms;
```

Comment:
We don't want this extern, static. Plus will get NULL initialization as well. I see NULL checks on it, so we probably want to make sure we initialize it to NULL IIUC.

Reply:
I added a `static`, thanks for the hint. Global variables are initialized with `NULL` implicitly, but I added an `= NULL` for clarity.